dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-to-fee-record-payment-groups.ts

```diff
@@ -0,0 +1,58 @@
+import {
+  FeeRecordPaymentEntityGroup,
+  calculateTotalCurrencyAndAmount,
+  getFeeRecordPaymentEntityGroupReconciliationData,
+  getFeeRecordPaymentEntityGroupStatus,
+} from '../../../../../helpers';
+import { mapPaymentEntityToPayment } from '../../../../../mapping/payment-mapper';
+import { mapFeeRecordEntityToFeeRecord } from '../../../../../mapping/fee-record-mapper';
+import { FeeRecordPaymentGroup } from '../../../../../types/utilisation-reports';
+
+/**
+ * Maps the fee record payment entity groups to the fee record payment groups
+ * @param feeRecordPaymentEntityGroups - The fee record payment entity groups
+ * @returns The fee record payment groups
+ */
+export const mapToFeeRecordPaymentGroups = async (feeRecordPaymentEntityGroups: FeeRecordPaymentEntityGroup[]): Promise<FeeRecordPaymentGroup[]> =>
+  await Promise.all(
+    feeRecordPaymentEntityGroups.map(async (group) => {
+      const groupStatus = getFeeRecordPaymentEntityGroupStatus(group);
+      const groupReconciliationData = await getFeeRecordPaymentEntityGroupReconciliationData(group);
```

Comment:
This seems quite expensive - how many typical payment groups will be passed here?

I am thinking perhaps we should do a single call to get multiple group's reconciliation data, instead of multiple single calls inside this mapping function; Lots going on in `getFeeRecordPaymentEntityGroupReconciliationData`

Reply:
In terms of performance, the `Promise.all` means this operation isn't actually very expensive in real terms.

The current solution has been designed to reach a compromise between maintainability, readability and performance. Whilst there may be a marginal gain in performance to have a single `mongoDbClient` call, I think the loss of readability and maintainability would not make up the difference

Reply:
How many payment groups will be passed here on average, is there a maximum?

Reply:
Realistically, about 10-20 at absolute max. In real terms, you're maybe looking at between 1 and 5

Reply:
I think this could be improved, but feel this is OK for a first iteration.

If some performance results arise we can look at optimising.